src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -63,6 +65,69 @@ private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
             }
         }
 
+        private void Workspace_WorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            switch (e.Kind)
+            {
+            case WorkspaceChangeKind.SolutionRemoved:
+            case WorkspaceChangeKind.SolutionCleared:
+                ClearOpenFileCache();
+                break;
+
+            case WorkspaceChangeKind.ProjectRemoved:
+                ClearOpenFileCache(e.ProjectId);
+                break;
+
+            default:
+                break;
+            }
+        }
+
+        private void ClearOpenFileCache(ProjectId projectId = null)
+        {
+            lock (_gate)
+            {
+                var itemReleased = false;
+                foreach (var (documentId, contextTask) in _openDocumentContexts)
+                {
+                    if (projectId is null || documentId.ProjectId == projectId)
+                    {
+                        itemReleased = true;
+                        ReleaseContext_NoLock(contextTask);
+                    }
+                }
+
+                // If any items were released due to the Clear operation, we need to remove them from the map.
+                if (itemReleased)
+                {
+                    if (projectId is null)
+                    {
+                        _openDocumentContexts.Clear();
+                    }
+                    else
+                    {
+                        var itemsToRemove = _openDocumentContexts.Keys.Where(key => key.ProjectId == projectId).ToList();
+                        foreach (var documentId in itemsToRemove)
+                        {
+                            _openDocumentContexts.Remove(documentId);
+                        }
+                    }
+                }
+            }
+        }
+
+        private void ReleaseContext_NoLock(Task<ICodingConventionContext> contextTask)
+        {
+            Debug.Assert(Monitor.IsEntered(_gate));
```

Comment:
Slick. Not sure why we don't do this in more places.

Reply:
👍 Now we know 😄